README.md
# dynamodb-session-web

An implementation of a "web" session using DynamoDB as backend storage. This project has the following goals:
* Focus on core session handling concerns, rather than specific Python frameworks.
* Follow [OWASP Session Management](https://cheatsheetseries.owasp.org/cheatsheets/Session_Management_Cheat_Sheet.html) 
best practices by default. Specifically, best practices for:
  * Session ID
    - [X] Length, value, and entropy: ID is a 32-byte secure random number. 
    - [X] Strict session management: ID generation can only occur within the framework.
    - [X] Treat as sensitive: A "loggable session ID" is provided as a property in order to make logging the ID easier (if desired).
  * Timeouts
    - [X] Absolute session timeout - default of 12 hours
    - [X] Idle session timeout - default of 2 hours
    - [X] Manual session timeout - i.e. there's delete/clear support

## Usage

Requires a DynamoDB table named `app_session` (can be changed in settings)

### Create session table

```shell
aws dynamodb create-table \
    --attribute-definitions \
        AttributeName=id,AttributeType=S \
    --key-schema "AttributeName=id,KeyType=HASH" \
    --provisioned-throughput "ReadCapacityUnits=5,WriteCapacityUnits=5" \
    --table-name app_session 
```

### Default Example
```python
from dynamodb_session_web import SessionManager

session = SessionManager()

# Create a new session object, get the ID for later use
initial_data = session.create()
session_id = initial_data.session_id

initial_data['foo'] = 'bar'
initial_data['one'] = 1
session.save(initial_data)

print(session_id)
#> 'WaHnSSou4d5Rq0k11vFGafe4sjMrkwiVhNziIWLLwMc'
print(initial_data.loggable_session_id)
#> '517286da2682be08dc9975612dc86d65487f0990906656f631d419e64dcda6f41f5e0529c290663be315524a0b35777645e0e827d2e982a048b5e2b4bba4e02b'

loaded_data = session.load(session_id)
print(loaded_data['foo'])
#> 'bar'
print(loaded_data['one'])
#> 1

session.clear(session_id)
```

### Configurable Timeout and NullSession Response
```python
from time import sleep
from dynamodb_session_web import SessionManager

session = SessionManager()

# Create a new session object, get the ID for later use
initial_data = session.create()
initial_data.idle_timeout_seconds = 30
initial_data.absolute_timeout_seconds = 30
session_id = initial_data.session_id

initial_data['foo'] = 'bar'
session.save(initial_data)

sleep(35)

loaded_data = session.load(session_id)
print(loaded_data)
#> <dynamodb_session_web.NullSessionInstance object at 0x109a7da30>
```


### Custom Data Class Example
```python
import json
from dataclasses import asdict, dataclass

from dynamodb_session_web import SessionInstanceBase, SessionManager

@dataclass
class MySession(SessionInstanceBase):
    fruit: str = ''
    color: str = ''

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

    def deserialize(self, data):
        data_dict = json.loads(data)
        self.fruit = data_dict['fruit']
        self.color = data_dict['color']

    def serialize(self):
        return json.dumps(asdict(self))

session = SessionManager(MySession)

# Create a new session object, get the ID for later use
initial_data = session.create()
session_id = initial_data.session_id

initial_data.fruit = 'apple'
initial_data.color = 'red'
session.save(initial_data)

loaded_data = session.load(session_id)
print(loaded_data.fruit)
#> 'apple'
print(loaded_data.color)
#> 'red'

session.clear(session_id)
```

## Configuration

Several behaviors can be configured at the Session Manager level:
* Custom data class; must provide serialization and deserialization methods (see examples). Defaults to a dictionary.
* Session ID length. Defaults to 32 bytes.
* Table name. Defaults to `app_session`
* DynamoDB URL. Defaults to None (i.e. Boto3 logic).
* Idle session timeout (in seconds). Defaults to 7200 seconds (2 hours).
* Absolute session timeout (in seconds). Defaults to 43200 seconds (12 hours).

```python
from dynamodb_session_web import SessionInstanceBase, SessionManager

class MyCustomDataClass(SessionInstanceBase):
    def deserialize(self, data: str):
        pass

    def serialize(self) -> str:
        pass

SessionManager(
    MyCustomDataClass,
    sid_byte_length=128,
    table_name='my-dynamodb-table',
    endpoint_url='http://localhost:8000',
    region_name='us-east-1',
    idle_timeout_seconds=300,
    absolute_timeout_seconds=3600,
)
```

Additionally, individual session instances can have their own idle and absolute timeouts, specified in seconds:

```python
from dynamodb_session_web import SessionManager

session = SessionManager()

instance = session.create()
instance.idle_timeout_seconds = 30
instance.absolute_timeout_seconds = 30
```

## Development

### Prerequisites:

* Docker

### Tests

[install `poetry`](https://python-poetry.org), then run `poetry install` to install the dependencies.

To run tests, run `poetry run pytest`

The integration tests will use the `docker-compose.yml` file to create a local DynamoDB instance.

## Useful Things

### Get a record from local DynamoDB instance

```shell
export sid=Jh4f1zvVp9n-YaDbkpZ0Vtru6iCXnERZv40q_ocZ7BA
aws dynamodb query --table-name app_session --endpoint-url http://localhost:8000 \
  --key-condition-expression 'id = :id' \
  --expression-attribute-values '{ ":id": {"S": "'$sid'" }}'
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.2.9](https://github.com/JCapriotti/dynamodb-session-web/tree/v0.2.9) - 2022-11-13

### Updated

- Allow package to be installed on Python >= 3.7; previously was >= 3.9, for no reason.
- Improved `tox` file for quicker runs and conciseness.
- Added explicit `Optional` to some method arguments; where before it was implicit.

## [0.2.8](https://github.com/JCapriotti/dynamodb-session-web/tree/v0.2.8) - 2022-09-09

### Added

- Allow a `region_name` parameter to be used when creating SessionManager instance, 
  which is passed to the internal `boto3` resource.

## Older

Nothing significant to mention


tests/helpers.py
from datetime import datetime
from unittest.mock import Mock

import boto3
from dynamodb_session_web import SessionManager, SessionDictInstance

TABLE_NAME = 'app_session'
LOCAL_ENDPOINT = 'http://localhost:8000'


def create_test_session(**kwargs) -> SessionManager[SessionDictInstance]:
    """
    Creates a SessionCore object configured for integration testing
    """
    return SessionManager(SessionDictInstance, endpoint_url=LOCAL_ENDPOINT, **kwargs)


def get_dynamo_resource():
    return boto3.resource('dynamodb', endpoint_url=LOCAL_ENDPOINT)


def get_dynamo_record(key):
    dynamodb = get_dynamo_resource()
    table = dynamodb.Table(TABLE_NAME)

    response = table.get_item(Key={'id': key})
    return response.get('Item', None)


def mock_current_datetime(mocker, val: datetime):
    mocker.patch('dynamodb_session_web._session.current_datetime', Mock(return_value=val))


tests/utility.py
from datetime import datetime
from random import randrange
from unittest.mock import Mock

import boto3
from dynamodb_session_web import SessionManager, SessionDictInstance

TABLE_NAME = 'app_session'
LOCAL_ENDPOINT = 'http://localhost:8000'
LOCAL_REGION_NAME = 'us-east-1'


def create_session_manager(**kwargs) -> SessionManager[SessionDictInstance]:
    """
    Creates a SessionCore object configured for integration testing
    """
    return SessionManager(SessionDictInstance, endpoint_url=LOCAL_ENDPOINT, region_name=LOCAL_REGION_NAME, **kwargs)


def get_dynamo_resource():
    return boto3.resource('dynamodb', endpoint_url=LOCAL_ENDPOINT, region_name=LOCAL_REGION_NAME)


def get_dynamo_record(key):
    dynamodb = get_dynamo_resource()
    table = dynamodb.Table(TABLE_NAME)

    response = table.get_item(Key={'id': key})
    return response.get('Item', None)


def mock_current_datetime(mocker, val: datetime):
    mocker.patch('dynamodb_session_web._session.current_datetime', Mock(return_value=val))


def str_param() -> str:
    return str(randrange(100000))


tests/conftest.py
import botocore
import pytest
from pytest_mock import MockerFixture
from dynamodb_session_web import SessionManager
from .utility import get_dynamo_resource, TABLE_NAME


@pytest.fixture(scope='function')
def dynamodb_table(docker_services):  # pylint: disable=unused-argument
    dynamodb = get_dynamo_resource()

    # Remove table (if it exists)
    # noinspection PyUnresolvedReferences
    try:
        table = dynamodb.Table(TABLE_NAME)
        table.delete()
    except botocore.exceptions.ClientError:
        pass

    # Create the DynamoDB table.
    table = dynamodb.create_table(
        TableName=TABLE_NAME,
        KeySchema=[{
            'AttributeName': 'id',
            'KeyType': 'HASH'
        }],
        AttributeDefinitions=[{
            'AttributeName': 'id',
            'AttributeType': 'S'
        }],
        ProvisionedThroughput={
            'ReadCapacityUnits': 5,
            'WriteCapacityUnits': 5
        }
    )

    # Wait until the table exists.
    table.meta.client.get_waiter('table_exists').wait(TableName=TABLE_NAME)
    yield
    table.delete()


@pytest.fixture
def mock_dynamo_set(mocker: MockerFixture):
    return mocker.patch.object(SessionManager, '_dynamo_set')


@pytest.fixture
def mock_dynamo_get(mocker: MockerFixture):
    return mocker.patch.object(SessionManager, '_dynamo_get')


dynamodb_session_web/_session.py
import hashlib
import json
from abc import ABC, abstractmethod
from datetime import datetime, timezone
from secrets import token_urlsafe
from typing import Any, Generic, List, NamedTuple, Optional, Type, TypeVar, overload

import boto3
from itsdangerous import BadSignature, Signer

from .exceptions import InvalidSessionIdError, SessionNotFoundError


DEFAULT_IDLE_TIMEOUT = 7200  # two hours
DEFAULT_ABSOLUTE_TIMEOUT = 43200  # twelve hours
DEFAULT_TABLE = 'app_session'
DEFAULT_SESSION_ID_BYTES = 32

T = TypeVar('T', bound='SessionInstanceBase')  # pylint: disable=invalid-name


class DynamoData(NamedTuple):
    data: Any
    idle_timeout: int
    absolute_timeout: int
    created: str


def create_session_id(byte_length: int, keys: List[str]) -> str:
    if len(keys) > 0:
        signer = Signer(keys)
        return signer.sign(token_urlsafe(byte_length)).decode('utf-8')
    return token_urlsafe(byte_length)


def validate_session_id(session_id, keys: List[str]) -> None:
    if not session_id:
        raise SessionNotFoundError(loggable_session_id(''))

    if len(keys) > 0:
        signer = Signer(keys)
        signer.unsign(session_id).decode('utf-8')


def current_datetime(datetime_value: Optional[datetime] = None) -> datetime:
    if datetime_value is None:
        datetime_value = datetime.now(tz=timezone.utc)
    return datetime_value


def current_timestamp(datetime_value: Optional[datetime] = None) -> int:
    return int(current_datetime(datetime_value).timestamp())


def expiration_datetime(idle_timeout: int, absolute_timeout: int, created: str, accessed: str) -> int:
    created_dt = datetime.fromisoformat(created)
    if created_dt.tzname() != 'UTC':
        raise ValueError("'created' must be UTC ")

    accessed_dt = datetime.fromisoformat(accessed)
    if accessed_dt.tzname() != 'UTC':
        raise ValueError("'accessed' must be UTC ")

    absolute_expiration = absolute_timeout + int(created_dt.timestamp())
    idle_expiration = idle_timeout + int(accessed_dt.timestamp())

    return min(absolute_expiration, idle_expiration)


def loggable_session_id(sid: str) -> str:
    return hashlib.sha512(sid.encode()).hexdigest()


class SessionInstanceBase(ABC):
    def __init__(self, *,
                 session_id: str = '',
                 idle_timeout_seconds: int = DEFAULT_IDLE_TIMEOUT,
                 absolute_timeout_seconds: int = DEFAULT_ABSOLUTE_TIMEOUT,
                 created: Optional[datetime] = None):
        self.session_id = session_id
        self.idle_timeout_seconds = int(idle_timeout_seconds)
        self.absolute_timeout_seconds = int(absolute_timeout_seconds)
        self.created = current_datetime() if created is None else created

    @abstractmethod
    def deserialize(self, data: str):
        pass

    @abstractmethod
    def serialize(self) -> str:
        pass

    @property
    def loggable_session_id(self):
        return loggable_session_id(self.session_id)


class SessionDictInstance(SessionInstanceBase, dict):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)

    def deserialize(self, data: str):
        self.update(json.loads(data))

    def serialize(self) -> str:
        return json.dumps(self)


class NullSessionInstance(SessionInstanceBase):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)

    def deserialize(self, data):
        pass

    def serialize(self):
        pass


class SessionManager(Generic[T]):  # pylint: disable=too-many-instance-attributes
    _boto_client = None
    _data_type: Type[T]
    _idle_timeout: int = DEFAULT_IDLE_TIMEOUT
    _absolute_timeout: int = DEFAULT_ABSOLUTE_TIMEOUT
    _sid_keys: List[str] = []

    null_session_class: Type[SessionInstanceBase] = NullSessionInstance

    @overload
    def __init__(self: 'SessionManager[SessionDictInstance]', **kwargs) -> None:  # pragma: no cover
        self.sid_byte_length = kwargs.get('sid_byte_length', DEFAULT_SESSION_ID_BYTES)
        self.table_name = kwargs.get('table_name', DEFAULT_TABLE)
        self.endpoint_url = kwargs.get('endpoint_url', None)
        self.region_name = kwargs.get('region_name', None)
        self._idle_timeout = kwargs.get('idle_timeout', DEFAULT_IDLE_TIMEOUT)
        self._absolute_timeout = kwargs.get('absolute_timeout', DEFAULT_ABSOLUTE_TIMEOUT)
        self._sid_keys = kwargs.get('sid_keys', [])
        self._bad_session_id_raises = kwargs.get('bad_session_id_raises', False)
        self._data_type = SessionDictInstance

    @overload
    def __init__(self: 'SessionManager[T]', data_type: Type[T], **kwargs) -> None:  # pragma: no cover
        self.sid_byte_length = kwargs.get('sid_byte_length', DEFAULT_SESSION_ID_BYTES)
        self.table_name = kwargs.get('table_name', DEFAULT_TABLE)
        self.endpoint_url = kwargs.get('endpoint_url', None)
        self.region_name = kwargs.get('region_name', None)
        self._idle_timeout = kwargs.get('idle_timeout', DEFAULT_IDLE_TIMEOUT)
        self._absolute_timeout = kwargs.get('absolute_timeout', DEFAULT_ABSOLUTE_TIMEOUT)
        self._sid_keys = kwargs.get('sid_keys', [])
        self._bad_session_id_raises = kwargs.get('bad_session_id_raises', False)
        self._data_type = data_type

    def __init__(self, data_type: Type[T] = SessionDictInstance, **kwargs) -> None:  # type: ignore
        """ Creates a new SessionManager instance.

        :param data_type: The data type to use for session instances. Defaults to SessionDictInstance.
        :param kwargs: Supported keyword args:
            sid_byte_length - The session id length in bytes. Defaults to 32.
            table_name - The DynamoDB table name. Defaults to app_session.
            endpoint_url - The DynamoDB URL. Defaults to None.
            region_name - The DynamoDB Region. Defaults to None.
            idle_timeout - The timeout used to expire idle sessions. Defaults to 7200 seconds.
            absolute_timeout - The timeout used for absolute session expiration. Defaults to 43200 seconds.
        """
        self.sid_byte_length = kwargs.get('sid_byte_length', DEFAULT_SESSION_ID_BYTES)
        self.table_name = kwargs.get('table_name', DEFAULT_TABLE)
        self.endpoint_url = kwargs.get('endpoint_url', None)
        self.region_name = kwargs.get('region_name', None)
        self._idle_timeout = kwargs.get('idle_timeout_seconds', DEFAULT_IDLE_TIMEOUT)
        self._absolute_timeout = kwargs.get('absolute_timeout_seconds', DEFAULT_ABSOLUTE_TIMEOUT)
        self._sid_keys = kwargs.get('sid_keys', [])
        self._bad_session_id_raises = kwargs.get('bad_session_id_raises', False)
        self._data_type = data_type

    def create(self, *,
               idle_timeout_seconds: Optional[int] = None,
               absolute_timeout_seconds: Optional[int] = None) -> T:
        """ Creates a session instance. At this point, no session record is persisted.

        :param idle_timeout_seconds: Idle timeout specific to this session instance.
        Defaults to session manager's value.

        :param absolute_timeout_seconds: Absolute timeout specific to this session instance.
        Defaults to session manager's value.

        :return: A new session instance
        """
        idle = self._idle_timeout if idle_timeout_seconds is None else idle_timeout_seconds
        absolute = self._absolute_timeout if absolute_timeout_seconds is None else absolute_timeout_seconds
        return self._data_type(session_id=create_session_id(self.sid_byte_length, self._sid_keys),
                               idle_timeout_seconds=idle,
                               absolute_timeout_seconds=absolute)

    def create_and_save(self, *,
                        idle_timeout_seconds: Optional[int] = None,
                        absolute_timeout_seconds: Optional[int] = None) -> T:
        """ Creates a session instance, and persists it in the database.

        :param idle_timeout_seconds: Idle timeout specific to this session instance.
        Defaults to session manager's value.

        :param absolute_timeout_seconds: Absolute timeout specific to this session instance.
        Defaults to session manager's value.

        :return: A new session instance
        """
        idle = self._idle_timeout if idle_timeout_seconds is None else idle_timeout_seconds
        absolute = self._absolute_timeout if absolute_timeout_seconds is None else absolute_timeout_seconds
        sid = create_session_id(self.sid_byte_length, self._sid_keys)
        session_data_object = self._data_type(session_id=sid,
                                              idle_timeout_seconds=idle,
                                              absolute_timeout_seconds=absolute)

        dynamo_data = DynamoData(session_data_object.serialize(),
                                 session_data_object.idle_timeout_seconds,
                                 session_data_object.absolute_timeout_seconds,
                                 current_datetime().isoformat())
        self._dynamo_set(dynamo_data, sid, modified=True)
        return session_data_object

    def load(self, session_id) -> T:
        data = None
        try:
            validate_session_id(session_id, self._sid_keys)
        except BadSignature as exc:
            if self._bad_session_id_raises:
                raise InvalidSessionIdError(loggable_session_id(session_id)) from exc
        else:
            data = self._perform_get(session_id)
            if data is None and self._bad_session_id_raises:
                raise SessionNotFoundError(loggable_session_id(session_id))

        if data is None:
            return self.null_session_class(session_id=session_id)  # type: ignore

        self._dynamo_set(data, session_id, modified=False)
        session_object = self._data_type(session_id=session_id,
                                         idle_timeout_seconds=data.idle_timeout,
                                         absolute_timeout_seconds=data.absolute_timeout,
                                         created=datetime.fromisoformat(data.created))
        session_object.deserialize(data.data)
        return session_object

    def save(self, data: T):
        dynamo_data = DynamoData(data.serialize(),
                                 data.idle_timeout_seconds,
                                 data.absolute_timeout_seconds,
                                 data.created.isoformat())
        self._dynamo_set(dynamo_data, data.session_id, modified=True)

    def clear(self, session_id):
        self._dynamo_remove(session_id)

    def _dynamo_remove(self, session_id):
        self.boto_client().delete_item(
            TableName=self.table_name,
            Key={'id': {'S': session_id}})

    def _perform_get(self, session_id) -> Optional[DynamoData]:
        return self._dynamo_get(session_id)

    def _dynamo_get(self, session_id) -> Optional[DynamoData]:
        res = self.boto_client().query(TableName=self.table_name,
                                       ExpressionAttributeValues={
                                           ':sid': {
                                               'S': session_id,
                                           },
                                           ':now': {
                                               'N': str(current_timestamp())
                                           },
                                       },
                                       KeyConditionExpression='id = :sid',
                                       FilterExpression='expires > :now',
                                       ConsistentRead=True)
        if res.get('Items') and len(res.get('Items')) == 1:
            idle_timeout = int(res.get('Items')[0]['idle_timeout'].get('N', self._idle_timeout))
            absolute_timeout = int(res.get('Items')[0]['absolute_timeout'].get('N', self._absolute_timeout))
            created: str = res.get('Items')[0]['created'].get('S', current_datetime().isoformat())

            if res.get('Items')[0]['data']:
                data = res.get('Items')[0]['data']
                return DynamoData(data.get('S', '{}'), idle_timeout, absolute_timeout, created)

        return None

    def _dynamo_set(self, data: DynamoData, session_id, modified):
        current_dt = current_datetime().isoformat()
        fields = {
            'accessed': {'S': current_dt},
            'expires': {'N': str(
                expiration_datetime(data.idle_timeout, data.absolute_timeout, data.created, current_dt)
            )},
            'created': {'S': data.created},
        }

        if modified:
            fields['idle_timeout'] = {'N': str(data.idle_timeout)}
            fields['absolute_timeout'] = {'N': str(data.absolute_timeout)}
            fields['data'] = {'S': data.data}

        attr_names = {}
        attr_values = {}
        update_expression = []
        for key, val in fields.items():
            attr = f'#attr_{key}'
            token = f':{key}'
            update_expression.append(f'{attr} = {token}')
            attr_values[token] = val
            attr_names[attr] = key

        self.boto_client().update_item(TableName=self.table_name,
                                       Key={'id': {'S': session_id}},
                                       ExpressionAttributeNames=attr_names,
                                       ExpressionAttributeValues=attr_values,
                                       UpdateExpression=f'SET {", ".join(update_expression)}',
                                       ReturnValues='NONE')

    def boto_client(self):
        if self._boto_client is None:
            boto_client_kwargs = {}
            if self.endpoint_url is not None:
                boto_client_kwargs["endpoint_url"] = self.endpoint_url
            if self.region_name is not None:
                boto_client_kwargs["region_name"] = self.region_name
            self._boto_client = boto3.client('dynamodb', **boto_client_kwargs)
        return self._boto_client


examples.py
import json
from dataclasses import asdict, dataclass
from time import sleep

from dynamodb_session_web import SessionManager, SessionDictInstance, SessionInstanceBase

LOCAL_ENDPOINT = 'http://localhost:8000'
TABLE_NAME = 'app_session'


def recreate_database():
    import boto3
    import botocore
    dynamodb = boto3.resource('dynamodb', endpoint_url=LOCAL_ENDPOINT)

    # Remove table (if it exists)
    try:
        table = dynamodb.Table(TABLE_NAME)
        table.delete()
    except botocore.exceptions.ClientError:
        pass

    # Create the DynamoDB table.
    return dynamodb.create_table(
        TableName=TABLE_NAME,
        KeySchema=[{
            'AttributeName': 'id',
            'KeyType': 'HASH'
        }],
        AttributeDefinitions=[{
            'AttributeName': 'id',
            'AttributeType': 'S'
        }],
        ProvisionedThroughput={
            'ReadCapacityUnits': 5,
            'WriteCapacityUnits': 5
        }
    )


def print_table_data(table):
    from decimal import Decimal

    class DecimalEncoder(json.JSONEncoder):
        def default(self, obj):
            if isinstance(obj, Decimal):
                return str(obj)
            return json.JSONEncoder.default(self, obj)

    response = table.scan(
        Select='ALL_ATTRIBUTES',
        ConsistentRead=True
    )
    if len(response.get('Items', [])) > 0:
        print(json.dumps(response['Items'], indent=2, cls=DecimalEncoder))
    else:
        print('>>>> No Data in DynamoDB <<<<')


dynamo_table = recreate_database()


@dataclass
class MySession(SessionInstanceBase):
    fruit: str = ''
    color: str = ''

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

    def deserialize(self, data):
        data_dict = json.loads(data)
        self.fruit = data_dict['fruit']
        self.color = data_dict['color']

    def serialize(self):
        return json.dumps(asdict(self))


def dataclass_example():
    """ Create a session manager using `MySession` as the data type """
    print('========== Data Class Example ==========')
    session = SessionManager(MySession, endpoint_url=LOCAL_ENDPOINT)

    # Create a new session object
    initial_data = session.create()
    session_id = initial_data.session_id

    initial_data.fruit = 'apple'
    initial_data.color = 'red'
    session.save(initial_data)

    print(f'Saved Data for {session_id}')
    print(f'The "loggable" session_id is: {initial_data.loggable_session_id}')
    print()
    print('Data from DynamoDB:')
    print_table_data(dynamo_table)
    print()

    loaded_data = session.load(session_id)
    print('Loaded data from Session record: ')
    print(f'Fruit: {loaded_data.fruit}')
    print(f'Color: {loaded_data.color}')
    print()

    session.clear(session_id)
    print('Cleared Data from DynamoDB:')
    print_table_data(dynamo_table)
    print()


def default_example():
    """ Create a session manager using default dictionary data type """
    print('========== Default Example ==========')
    session = SessionManager(endpoint_url=LOCAL_ENDPOINT)

    # Create a new session object
    initial_data = session.create()
    session_id = initial_data.session_id

    initial_data['foo'] = 'bar'
    initial_data['one'] = 1
    session.save(initial_data)

    print(f'Saved Data for {session_id}')
    print(f'The "loggable" session_id is: {initial_data.loggable_session_id}')
    print()
    print('Data from DynamoDB:')
    print_table_data(dynamo_table)
    print()

    loaded_data = session.load(session_id)
    print('Loaded data from Session record: ')
    print(f'foo: {loaded_data["foo"]}')
    print(f'one: {loaded_data["one"]}')
    print()

    session.clear(session_id)
    print('Cleared Data from DynamoDB:')
    print_table_data(dynamo_table)
    print()


def timeout_example():
    """ Example with extremely short timeout """
    print('========== Session Expiration Example ==========')
    session = SessionManager(SessionDictInstance, endpoint_url=LOCAL_ENDPOINT)

    # Create a new session object
    initial_data = session.create()
    initial_data.idle_timeout_seconds = 30
    initial_data.absolute_timeout_seconds = 30
    session_id = initial_data.session_id

    initial_data['foo'] = 'bar'
    session.save(initial_data)

    print(f'Saved Data for {session_id}')
    print(f'The "loggable" session_id is: {initial_data.loggable_session_id}')
    print()
    print('Data from DynamoDB:')
    print_table_data(dynamo_table)
    print()

    print('Sleeping for 35 seconds')
    sleep(35)

    loaded_data = session.load(session_id)
    print('Loaded data from Session record: ')
    print(loaded_data)
    print()

    print('Data from DynamoDB:')
    print_table_data(dynamo_table)


dataclass_example()
default_example()
timeout_example()
